File: electron-app/src/satClient.js
const { chromium } = require("playwright");
const path = require("path");
const fs = require("fs");

const SAT_URL = "https://portalcfdi.facturaelectronica.sat.gob.mx/";
const MAX_RETRIES = 2;

class SATClient {
  constructor(rutaCer, rutaKey, password, logFn = null) {
    this.rutaCer = rutaCer;
    this.rutaKey = rutaKey;
    this.password = password;
    this.logFn = logFn;
    this.browser = null;
    this.page = null;
  }

  log(msg) {
    if (this.logFn) this.logFn(msg);
  }

  _validateFiles() {
    if (!fs.existsSync(this.rutaCer)) {
      return "El archivo de certificado no existe: " + this.rutaCer;
    }
    if (!fs.existsSync(this.rutaKey)) {
      return "El archivo de llave no existe: " + this.rutaKey;
    }
    return null;
  }

  async _initBrowser() {
    const bundledBrowsers = path.join(process.resourcesPath || '', 'browsers');
    let hasChromium = false;
    try {
      if (fs.existsSync(bundledBrowsers)) {
        const entries = fs.readdirSync(bundledBrowsers);
        hasChromium = entries.some(e => e.startsWith('chromium'));
      }
    } catch (e) {}
    if (hasChromium) {
      process.env.PLAYWRIGHT_BROWSERS_PATH = bundledBrowsers;
    } else if (!process.env.PLAYWRIGHT_BROWSERS_PATH) {
      process.env.PLAYWRIGHT_BROWSERS_PATH = "0";
    }
    const launchOptions = { headless: false };
    if (process.platform === 'darwin') {
      launchOptions.channel = 'chrome';
    }
    this.browser = await chromium.launch(launchOptions);
    const context = await this.browser.newContext({
      acceptDownloads: true,
      viewport: { width: 1366, height: 768 },
    });
    context.setDefaultTimeout(0);
    this.page = await context.newPage();
    this.page.setDefaultTimeout(0);
    this.page.setDefaultNavigationTimeout(0);
  }

  async login() {
    const fileError = this._validateFiles();
    if (fileError) {
      this.log(fileError);
      return { ok: false, msg: fileError };
    }

    this.log("Iniciando navegador...");
    try {
      await this._initBrowser();
    } catch (e) {
      this.log(e.message);
      return { ok: false, msg: e.message };
    }
    const page = this.page;

    try {
      this.log("Abriendo portal SAT...");
      await page.goto(SAT_URL);
      await page.waitForLoadState("load");

      this.log("Haciendo clic en 'e.firma'...");
      const efirmaBtn = page.locator("#buttonFiel");
      await efirmaBtn.waitFor();
      await efirmaBtn.click();
      await page.waitForTimeout(2000);

      this.log(`Subiendo certificado: ${this.rutaCer}`);
      const [fc1] = await Promise.all([
        page.waitForEvent("filechooser"),
        page.locator("#btnCertificate").click(),
      ]);
      await fc1.setFiles(this.rutaCer);
      this.log("Certificado subido");

      this.log(`Subiendo llave: ${this.rutaKey}`);
      const [fc2] = await Promise.all([
        page.waitForEvent("filechooser"),
        page.locator("#btnPrivateKey").click(),
      ]);
      await fc2.setFiles(this.rutaKey);
      this.log("Llave subida");

      this.log("Ingresando password...");
      const pwInput = page.locator("input[type='password']").first();
      await pwInput.waitFor();
      await pwInput.fill(this.password);

      let satDialogMsg = null;
      page.once("dialog", dialog => {
        satDialogMsg = dialog.message();
        this.log(`Dialog del SAT: ${satDialogMsg}`);
        dialog.accept();
      });

      this.log("Enviando formulario de inicio de sesión...");
      await page.locator("#submit").waitFor({ state: "visible" });
      await page.waitForFunction(() => !document.getElementById("submit")?.disabled);
      await page.locator("#submit").click();

      this.log("Esperando autenticación...");
      await page.waitForTimeout(5000);

      if (satDialogMsg) {
        return { ok: false, msg: `El SAT indicó: ${satDialogMsg}` };
      }

      const currentUrl = page.url().toLowerCase();
      if (currentUrl.includes("error") || currentUrl.includes("bloqueado")
        || currentUrl.includes("suspender") || currentUrl.includes("mantenimiento")) {
        this.log("Error de autenticación - revisa certificados y password");
        return { ok: false, msg: "Error de autenticación en el SAT. Revisa tus certificados." };
      }

      try {
        const bodyText = await page.evaluate(() => document.body.innerText.toLowerCase());
        if (bodyText.includes("vencido") || bodyText.includes("revocado")) {
          return { ok: false, msg: "El certificado está vencido o revocado." };
        }
        if (bodyText.includes("no corresponde")) {
          return { ok: false, msg: "El archivo de llave no corresponde al certificado." };
        }
        if (bodyText.includes("contraseña") || bodyText.includes("password") || bodyText.includes("incorrecta")) {
          return { ok: false, msg: "Contraseña incorrecta." };
        }
      } catch (e) {
        // ignore errors reading body text
      }

      this.log("Inicio de sesión exitoso");
      return { ok: true, msg: "Login exitoso" };
    } catch (e) {
      const msg = e.message;
      this.log(`Error durante autenticación: ${msg}`);
      if (msg.includes("timeout") || msg.includes("Timeout") || msg.includes("net::ERR_")) {
        return { ok: false, msg: "Error de conexión con el SAT. Verifica tu conexión a internet." };
      }
      return { ok: false, msg: msg };
    }
  }

  async navigateToEmitidas() {
    return this._navigateToFacturas("emitidas");
  }

  async navigateToRecibidas() {
    const page = this.page;
    if (!page) return { ok: false, msg: "Navegador no inicializado" };
    try {
      this.log("Navegando a facturas recibidas...");
      const link = page.locator('a[title="Facturas Recibidas"]');
      await link.waitFor();
      await link.click();
      await page.waitForLoadState("load");
      this.log("Navegación exitosa a recibidas");
      return { ok: true };
    } catch (e) {
      return { ok: false, msg: `No se pudo navegar a facturas recibidas: ${e.message}` };
    }
  }

  async _navigateToFacturas(tipo) {
    const page = this.page;
    if (!page) return { ok: false, msg: "Navegador no inicializado" };

    try {
      this.log(`Navegando a facturas ${tipo}...`);
      const link = page.locator(`text=Facturas ${tipo}`).first();
      await link.waitFor();
      await link.click();
      await page.waitForLoadState("load");
      this.log("Navegación exitosa");
      return { ok: true };
    } catch {
      this.log("No se encontró el enlace directo. Intentando vía menú...");
      try {
        const consultas = page.locator("text=Consultas").first();
        await consultas.waitFor();
        await consultas.click();
        await page.waitForTimeout(1000);
        const subLink = page.locator(`text=Facturas ${tipo}`).first();
        await subLink.waitFor();
        await subLink.click();
        await page.waitForLoadState("load");
        return { ok: true };
      } catch (e) {
        return { ok: false, msg: `No se pudo navegar a facturas ${tipo}` };
      }
    }
  }

  async searchPeriodo(year, month) {
    const page = this.page;
    if (!page) return { ok: false, msg: "Navegador no inicializado" };

    try {
      await page.waitForTimeout(2000);

      this.log("Seleccionando filtro por fechas...");
      await page.locator("input#ctl00_MainContent_RdoFechas").click();
      this.log("Esperando respuesta del servidor...");
      await page.waitForTimeout(5000);

      this.log(`Seleccionando año ${year}...`);
      await page.locator("#DdlAnio").selectOption(String(year));

      this.log(`Seleccionando mes ${String(month).padStart(2, "0")}...`);
      await page.locator("#ctl00_MainContent_CldFecha_DdlMes").selectOption(String(month));

      await page.waitForTimeout(500);

      this.log("Haciendo clic en Buscar CFDI...");
      await page.locator("#ctl00_MainContent_BtnBusqueda").click();

      this.log("Esperando resultados de búsqueda...");
      await page.waitForTimeout(8000);

      const downloadBtns = page.locator("span.glyphicon-cloud-download, span[onclick*='AccionCfdi']");
      const total = await downloadBtns.count();
      this.log(`Facturas encontradas: ${total}`);

      return { ok: true, total };
    } catch (e) {
      return { ok: false, msg: e.message };
    }
  }

  async requestDownloadAll(downloadPath) {
    const page = this.page;
    if (!page) return { ok: false, msg: "Navegador no inicializado" };

    try {
      await page.locator("span.glyphicon-cloud-download[onclick]").first().waitFor({ timeout: 15000 }).catch(() => {});
      await page.waitForTimeout(2000);

      const urlParts = await page.locator("span.glyphicon-cloud-download[onclick]").evaluateAll(
        (els) => els.map((el) => {
          const onclick = el.getAttribute("onclick");
          const match = onclick.match(/AccionCfdi\('([^']+)'/);
          return match ? match[1] : null;
        }).filter(Boolean)
      );

      const total = urlParts.length;

      if (total === 0) {
        this.log("No hay facturas para descargar en este periodo");
        return { ok: true, downloaded: 0, failed: 0, total: 0 };
      }

      if (downloadPath) {
        fs.mkdirSync(downloadPath, { recursive: true });
      }

      this.log(`Facturas encontradas: ${total}`);

      const baseUrl = page.url();
      let downloaded = 0;
      let failed = 0;

      for (let i = 0; i < total; i++) {
        let success = false;
        for (let retry = 1; retry <= 2; retry++) {
          try {
            this.log(`Descargando ${i + 1}/${total}${retry > 1 ? ` (intento ${retry})` : ''}...`);
            const resolvedUrl = new URL(urlParts[i], baseUrl).href;
            const xml = await page.evaluate(async (url) => {
              const res = await fetch(url);
              return await res.text();
            }, resolvedUrl);

            if (!xml || xml.trim().length === 0) {
              throw new Error("Contenido XML vacío");
            }

            const uuidMatch = xml.match(/UUID=["']([a-fA-F0-9\-]{36})["']/);
            const uuid = uuidMatch ? uuidMatch[1] : null;
            const fileName = uuid ? `${uuid}.xml` : `cfdi_${i + 1}.xml`;
            const destDir = downloadPath || ".";
            const filePath = path.join(destDir, fileName);
            fs.writeFileSync(filePath, xml, "utf-8");
            this.log(`✓ ${i + 1}/${total} — ${filePath}`);
            downloaded++;
            success = true;
            break;
          } catch (e) {
            if (retry < 2) {
              this.log(`Reintentando ${i + 1}/${total}...`);
              await page.waitForTimeout(3000);
            } else {
              this.log(`✗ ${i + 1}/${total} — Error: ${e.message}`);
              failed++;
            }
          }
        }
        await page.waitForTimeout(2000);
      }

      const summary = `Periodo completado: ${downloaded} OK, ${failed} FALLIDAS`;
      if (failed > 0) {
        this.log(`⚠️  ${summary}`);
      } else {
        this.log(summary);
      }
      return { ok: true, downloaded, failed, total };
    } catch (e) {
      return { ok: false, msg: e.message };
    }
  }

  async retrieveDownloads(downloadPath) {
    const page = this.page;
    if (!page) return { ok: false, msg: "Navegador no inicializado" };

    try {
      this.log("Navegando a Recuperar descargas...");
      const recuperarLink = page.locator("text=Recuperar descargas").first();
      if ((await recuperarLink.count()) === 0) {
        const consultas = page.locator("text=Consultas").first();
        if ((await consultas.count()) > 0) {
          await consultas.click();
          await page.waitForTimeout(1000);
        }
      }
      const recuperar = page.locator("text=Recuperar descargas").first();
      if ((await recuperar.count()) > 0) {
        await recuperar.click();
        await page.waitForLoadState("load");
        await page.waitForTimeout(3000);
      }

      const downloadLinks = page.locator("a:has-text('Descargar'), button:has-text('Descargar')");
      const linkCount = await downloadLinks.count();

      if (linkCount === 0) {
        return { ok: false, msg: "No hay descargas disponibles para recuperar. Las solicitudes pueden tardar minutos en procesarse." };
      }

      let downloaded = 0;
      let failed = 0;

      for (let i = 0; i < linkCount; i++) {
        const link = downloadLinks.nth(i);
        try {
          const [download] = await Promise.all([
            page.waitForEvent("download"),
            link.click(),
          ]);

          const now = new Date();
          const dateStr = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`;
          const destDir = downloadPath || path.join(process.cwd(), "downloads", dateStr);
          fs.mkdirSync(destDir, { recursive: true });

          const fileName = download.suggestedFilename() || `descarga_${i}_${dateStr}.zip`;
          const filePath = path.join(destDir, fileName);
          await download.saveAs(filePath);
          this.log(`Descargado: ${fileName}`);
          downloaded++;
        } catch (e) {
          this.log(`Error al descargar: ${e.message}`);
          failed++;
        }
        await page.waitForTimeout(1000);
      }

      return { ok: true, downloaded, failed };
    } catch (e) {
      return { ok: false, msg: e.message };
    }
  }

  async downloadPeriodo(year, month, tipo, downloadPath) {
    const page = this.page;
    if (!page) return { ok: false, msg: "Navegador no inicializado" };

    const meses = [
      "Enero", "Febrero", "Marzo", "Abril", "Mayo", "Junio",
      "Julio", "Agosto", "Septiembre", "Octubre", "Noviembre", "Diciembre",
    ];
    const mesStr = meses[month - 1];

    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
      try {
        this.log(`[${attempt}/${MAX_RETRIES}] Periodo: ${mesStr} ${year} (${tipo})`);

        const searchResult = await this.searchPeriodo(year, month);
        if (!searchResult.ok) {
          if (attempt < MAX_RETRIES) {
            this.log("Reintentando búsqueda...");
            await page.waitForTimeout(3000);
            continue;
          }
          return { ok: false, msg: `Error en búsqueda: ${searchResult.msg}` };
        }

        const downloadResult = await this.requestDownloadAll(downloadPath);
        if (downloadResult.total > 0) {
          this.log(`Descargadas: ${downloadResult.downloaded}/${downloadResult.total}`);
        }
        if (downloadResult.failed > 0) {
          this.log(`${downloadResult.failed} fallidas`);
        }

        return {
          ok: true,
          msg: `${mesStr} ${year} - ${downloadResult.downloaded}/${downloadResult.total} descargadas`,
        };
      } catch (e) {
        this.log(`Error en intento ${attempt}: ${e.message}`);
        if (attempt < MAX_RETRIES) {
          this.log("Reintentando...");
          await page.waitForTimeout(3000);
        } else {
          return { ok: false, msg: `Error: ${e.message}` };
        }
      }
    }

    return { ok: false, msg: "Falló después de reintentos" };
  }

  async close() {
    try {
      if (this.browser) await this.browser.close();
    } catch {}
    this.page = null;
    this.browser = null;
  }
}

module.exports = { SATClient };

File: app/core/sat_client.py
import os
import time
from typing import Optional

from playwright.sync_api import sync_playwright, Page, Browser, TimeoutError as PWTimeout

SAT_URL = "https://portalcfdi.facturaelectronica.sat.gob.mx/"
MAX_RETRIES = 2

class SATClient:
    def __init__(self, ruta_cer: str, ruta_key: str, password: str,
                headless: bool = False, log_fn=None):
        self.ruta_cer = ruta_cer
        self.ruta_key = ruta_key
        self.password = password
        self.headless = headless
        self.log_fn = log_fn
        self._playwright = None
        self._browser: Optional[Browser] = None
        self._page: Optional[Page] = None

    def log(self, msg: str):
        if self.log_fn:
            self.log_fn(msg)

    def _ensure_browsers(self):
        from playwright._impl._driver import compute_driver_executable
        driver_exec = compute_driver_executable()
        import subprocess
        result = subprocess.run(
            [str(driver_exec), "install", "chromium"],
            capture_output=True, text=True, timeout=120,
        )
        if result.returncode != 0:
            self.log(f"Instalando Chromium para Playwright...")
            result = subprocess.run(
                [str(driver_exec), "install", "chromium"],
                capture_output=True, text=True, timeout=180,
            )
            if result.returncode != 0:
                self.log(f"Error al instalar Chromium: {result.stderr[:200]}")

    def _init_browser(self):
        self._ensure_browsers()
        self._playwright = sync_playwright().start()
        self._browser = self._playwright.chromium.launch(
            headless=self.headless,
        )
        context = self._browser.new_context(
            accept_downloads=True,
            viewport={"width": 1366, "height": 768},
        )
        self._page = context.new_page()

    def login(self) -> bool:
        self.log("Iniciando navegador...")
        self._init_browser()
        page = self._page

        try:
            self.log("Abriendo portal SAT...")
            page.goto(SAT_URL, timeout=30000)
            page.wait_for_load_state("networkidle")

            self.log("Haciendo clic en 'Iniciar sesión'...")
            iniciar_btn = page.locator("text=Iniciar sesión").first
            iniciar_btn.wait_for(timeout=15000)
            iniciar_btn.click()
            time.sleep(2)

            self.log("Seleccionando 'Certificado de sello digital'...")
            cert_option = page.locator("text=Certificado de sello digital").first
            cert_option.wait_for(timeout=10000)
            cert_option.click()
            time.sleep(1)

            self.log(f"Subiendo certificado: {self.ruta_cer}")
            cer_input = page.locator("input[type='file']").first
            cer_input.wait_for(timeout=10000)
            cer_input.set_input_files(self.ruta_cer)
            time.sleep(1)

            self.log(f"Subiendo llave: {self.ruta_key}")
            file_inputs = page.locator("input[type='file']")
            count = file_inputs.count()
            if count > 1:
                file_inputs.nth(1).set_input_files(self.ruta_key)
            else:
                page.wait_for_timeout(2000)
                key_inputs = page.locator("input[type='file']")
                if key_inputs.count() > 1:
                    key_inputs.nth(1).set_input_files(self.ruta_key)
                else:
                    self.log("Error: No se encontró el campo para la llave")
                    return False
            time.sleep(1)

            self.log("Ingresando password...")
            pw_input = page.locator("input[type='password']").first
            pw_input.wait_for(timeout=10000)
            pw_input.fill(self.password)

            self.log("Enviando formulario de inicio de sesión...")
            submit_btn = page.locator("button[type='submit']").first
            if not submit_btn.is_visible():
                submit_btn = page.locator("text=Ingresar").first
            submit_btn.click()

            self.log("Esperando autenticación...")
            page.wait_for_timeout(5000)

            current_url = page.url
            if "error" in current_url.lower():
                self.log("Error de autenticación - revisa certificados y password")
                return False

            self.log("Inicio de sesión exitoso")
            return True

        except PWTimeout:
            self.log("Tiempo de espera agotado durante la autenticación en SAT")
            return False
        except Exception as e:
            self.log(f"Error durante autenticación: {str(e)}")
            return False

    def navigate_to_emitidas(self) -> bool:
        return self._navigate_to_facturas("emitidas")

    def navigate_to_recibidas(self) -> bool:
        return self._navigate_to_facturas("recibidas")

    def _navigate_to_facturas(self, tipo: str) -> bool:
        page = self._page
        if not page:
            self.log("Navegador no inicializado.")
            return False

        try:
            self.log(f"Navegando a facturas {tipo}...")
            link = page.locator(f"text=Facturas {tipo}").first
            link.wait_for(timeout=15000)
            link.click()
            page.wait_for_load_state("networkidle")
            self.log("Navegación exitosa")
            return True
        except PWTimeout:
            self.log("No se encontró el enlace directo. Intentando vía menú...")
            try:
                consultas = page.locator("text=Consultas").first
                consultas.wait_for(timeout=5000)
                consultas.click()
                time.sleep(1)
                sub = page.locator(f"text=Facturas {tipo}").first
                sub.wait_for(timeout=5000)
                sub.click()
                page.wait_for_load_state("networkidle")
                return True
            except PWTimeout:
                self.log("No se pudo navegar a la sección de facturas")
                return False

    def download_periodo(self, year: int, month: int, tipo: str = "emitidas",
                        download_path: Optional[str] = None) -> dict:
        page = self._page
        if not page:
            return {"ok": False, "msg": "Navegador no inicializado"}

        for attempt in range(1, MAX_RETRIES + 1):
            try:
                meses = [
                    "Enero", "Febrero", "Marzo", "Abril", "Mayo", "Junio",
                    "Julio", "Agosto", "Septiembre", "Octubre", "Noviembre", "Diciembre",
                ]
                mes_str = meses[month - 1]

                self.log(f"[{attempt}/{MAX_RETRIES}] Periodo: {mes_str} {year} ({tipo})")

                if download_path:
                    os.makedirs(download_path, exist_ok=True)

                page.wait_for_timeout(2000)

                year_selector = page.locator("select").filter(has_text=page.locator("text=202")).first
                if year_selector.count() > 0:
                    year_selector.select_option(str(year))

                month_selector = page.locator("select").filter(has_text=page.locator(f"text={mes_str[:4]}")).first
                if month_selector.count() > 0:
                    month_selector.select_option(str(month))

                page.wait_for_timeout(1000)

                buscar_btn = page.locator("button:has-text('Buscar')").first
                if buscar_btn.count() == 0:
                    buscar_btn = page.locator("button:has-text('Consultar')").first
                if buscar_btn.count() == 0:
                    buscar_btn = page.locator("input[type='submit']").first

                if buscar_btn.count() > 0:
                    buscar_btn.click()
                    self.log("Buscando...")
                    page.wait_for_timeout(8000)
                else:
                    self.log("No se encontró botón de búsqueda, continuando...")

                self.log("Iniciando descarga...")
                page.wait_for_timeout(3000)
                self.log(f"Descarga completada para {mes_str} {year}")

                return {"ok": True, "msg": f"{mes_str} {year}"}

            except PWTimeout:
                self.log(f"Timeout en intento {attempt}")
                if attempt < MAX_RETRIES:
                    self.log("Reintentando...")
                    page.wait_for_timeout(3000)
                else:
                    return {"ok": False, "msg": f"Timeout {mes_str} {year}"}
            except Exception as e:
                self.log(f"Error en intento {attempt}: {str(e)}")
                if attempt < MAX_RETRIES:
                    self.log("Reintentando...")
                    page.wait_for_timeout(3000)
                else:
                    return {"ok": False, "msg": f"Error: {str(e)}"}

        return {"ok": False, "msg": "Falló después de reintentos"}

    def close(self):
        try:
            if self._browser:
                self._browser.close()
            if self._playwright:
                self._playwright.stop()
        except Exception:
            pass
        self._page = None
        self._browser = None
        self._playwright = None
